### PR TITLE
[addons] add architecture dependent platform strings (binary addon repo)

### PR DESCRIPTION
**Pull request by Rechi**
## Description
~~Adds a `platformdependent` tag to addon.xml which causes that the addon will be downloaded at `<ADDON-ID>/<ADDON-ID>-<VERSION>-<PLATFORM>.zip` instead of `<ADDON-ID>/<ADDON-ID>-<VERSION>.zip` The PLATFORM is the first match of the platform tag.
Additionally I added architecture dependant platform selectors to get the correct addon depending on the architecture Kodi was build for.~~
Adds `<PLATFROM>-<ARCH>`  filters for the platform tag in addon.xml. The download URL is determined by the `path` tag in the addons.xml (inserted by repository-generator). For our setup it is `<ADDON-ID>+<PLATFROM_TAG>/<ADDON-ID>-<VERSION>.zip`.

## How Has This Been Tested?
installed pvr.argustv on windows from my test repo

## Types of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by MartijnKaijser** [comment]
Will the test repo be dropped and integrated into our already existing addon repo list?

**Comment by Rechi** [comment]
That's the plan.

**Comment by Jalle19** [comment]
I think the word you're looking for is "dependent", not "dependant"?

**Comment by tamland** [comment]
>Adds a platformdependant tag to addon.xml which causes that the addon will be downloaded at <ADDON-ID>/<ADDON-ID>-<VERSION>-<PLATFORM>.zip instead of <ADDON-ID>/<ADDON-ID>-<VERSION>.zip The PLATFORM is the first match of the platform tag.

Since this depends on the platform tag I would rather add 'platformdependant' (or whatever) as an attribute to `<platform>` instead. It's more xml-y

File: xbmc/addons/AddonManager.cpp
Comment (tamland) [inline xbmc/addons/AddonManager.cpp]: Please follow the existing structure and keep metadata parsing in the `Factory` method. Even if you just need it for the path right now, there is no need to throw this info away.

File: cmake/scripts/common/AddonHelpers.cmake
Comment (wsnipex) [inline cmake/scripts/common/AddonHelpers.cmake]: shouldn't this be known from kodiconfig.cmake? if not, it'd be worth adding there

Comment (koying) [reply cmake/scripts/common/AddonHelpers.cmake]: Seems a bit hacky. And we have to split the arch for android as well. Linux also, if in-scope.
We probably need a more structured way to get the arch/cpu here...

Comment (Rechi) [reply cmake/scripts/common/AddonHelpers.cmake]: This was only hacked in to show it works, needs to be done properly for Windows, Android, iOS and OS X.
For Linux we still use the ppa, and distros like LibreELEC should define their own naming if they want a binary addon repo.

Comment (koying) [reply cmake/scripts/common/AddonHelpers.cmake]: Ok. So, If I got you right, we might as well do a mass rename in jenkins, which is already per arch, can we?

**Comment by koying** [comment]
Isn't the `platformdependant` tag redundant with the `platform` tag? 
Ie do you see a use case where there would be a platform tag with `platformdependant=false` (or "platformdependant=true" without "platform"?)

**Comment by Rechi** [comment]
`platform` filter tag is used by some python addons too (e.g. script.kodi.android.update, ) and I don't want to change the download path for those addons. There are also some addons which list multiple platforms, but do not need different code (one universal package which fits for all listed platforms).
Maybe there is a better name instead of `platformdependent` as it actually means the correct package for this platform and architecture is needed (there isn't one universal package which fits for all listed platforms).

**Comment by koying** [comment]
Right.
As it' basically a repository setting, couldn't we put platformdependent at that level?
Just trying to find a solution to not have to revisit every single addon.xml

**Comment by Rechi** [comment]
Binary repo won't be a separate repo, it has to go into https://mirrors.kodi.tv/addons/*/addons.xml.
Otherwise a binary repo wouldn't need any special cases inside Kodi, but every Kodi instance would send two requests to our mirrors, which isn't acceptable.

**Comment by koying** [comment]
OK.
What do you think of https://github.com/koying/xbmc/commit/7b288463200fcc6595e50ffcd7792574a9c21420 for android?

**Comment by wsnipex** [comment]
s/platformdependent/archspecific?

**Comment by koying** [comment]
Used https://github.com/koying/xbmc/commit/855fae32f537d386d1fa93a42f1e526bccd50f1a to progress.
If binary addon and platform specific -> use platform tagging

**Comment by koying** [comment]
There is some decision to take regarding the <platform> tag *inside* the plugins.
Currently, for android, `<platform>android</platform>` is kept, so not arch dependent. Only the package filename is altered. That also means that the repo addons.xml does not contain the same as the the plugin addon.xml.

If we want something like `<platform>android-armv7</platform>`, which seem more elegant to me, we  have to decide if we use `library_android=` or `library_android-armv7=` in addon.xml.

**Comment by tamland** [comment]
When I think about maybe it's easier and better to just include the download location in addons.xml server side and have the client use that. That way you don't need this extra flag just for the client to figure it where it should be located. What do you think @Rechi?

**Comment by MartijnKaijser** [comment]
It should be added to addon.xml at compile time because people will also download the zips separately and not just from repo

**Comment by Rechi** [comment]
> Currently, for android, `<platform>android</platform>` is kept, so not arch dependent. Only the package filename is altered. That also means that the repo addons.xml does not contain the same as the the plugin addon.xml.

Actually I meant that the addon,xml files in the repo have to be changed, repository-generator shouldn't do that.

> If we want something like `<platform>android-armv7</platform>`, which seem more elegant to me, we  have to decide if we use `library_android=` or `library_android-armv7=` in addon.xml.

Can't we switch to `library=` instead of `library_@PLATFORM@` or is this still needed for something?

> When I think about maybe it's easier and better to just include the download location in addons.xml server side and have the client use that.

Would also have the advantage to not change behaviour for other binary repos (e.g. LibreELEC)

> It should be added to addon.xml at compile time because people will also download the zips separately and not just from repo

**Comment by MartijnKaijser** [comment]
Exactly what you said. It should not be done by the repo generator

**Comment by koying** [comment]
> Can't we switch to library= instead of library_@PLATFORM@ or is this still needed for something?

I assume an original thought was to bundle all platform-dependent libraries in a single package?
If we definitely abandon this idea, then indeed I don't see the point of having `library_<PLATFORM>`

**Comment by tamland** [comment]
>Exactly what you said. It should not be done by the repo generator

Why? That doesn't make any sense. I'm not talking about modifying the addon.xml in the package, but adding information to the addons.xml file. That's already being done for other things.

**Comment by MartijnKaijser** [comment]
It doesn't make sense to let the repo generator define what platform it should be made available for. As I already said we should not depend on a repo generator to mandate the platform a zipped add-on should be compatible with. this is must do inside the initial addon.xml file and the repo generator can take that info.

**Comment by wsnipex** [comment]
addon manager uses library_@PLATFORM@ to decide which library to load

**Comment by koying** [comment]
> addon manager uses library_@PLATFORM@ to decide which library to load

Sure. Point is that if 1 package = 1 arch, there is no need to have the library attribute per arch as well, so https://github.com/xbmc/xbmc/blob/master/xbmc/addons/AddonManager.cpp#L1092 could go

**Comment by MartijnKaijser** [comment]
So in the end we still need the following
https://github.com/xbmc/xbmc/pull/5836
https://github.com/xbmc/xbmc/pull/5837

**Comment by tamland** [comment]
> It doesn't make sense to let the repo generator define what platform it should be made available for. As I already said we should not depend on a repo generator to mandate the platform a zipped add-on should be compatible with. this is must do inside the initial addon.xml file and the repo generator can take that info.
>
>Edit:
linux will probably not have a binary repo to begin with and uses a PPA, so the addon.xml should still have the exact info needed to say if it can be installed/loaded or not.
Also installing a binary addon zip file bypasses the addons.xml and as such only doing it in there makes it pointless.

You completely misunderstood. That's doesn't have anything to do with what we're talking about. My suggesting was about explicitly saying where the corresponding zip file is located instead of having it implicitly determined client side by some conversion. Nothing about compatibility is changed.

**Comment by tamland** [comment]
>>    addon manager uses library_@PLATFORM@ to decide which library to load
>
>Sure. Point is that if 1 package = 1 arch, there is no need to have the library attribute per arch as well, so https://github.com/xbmc/xbmc/blob/master/xbmc/addons/AddonManager.cpp#L1092 could go

Would be nice to not destroy the ability to make a package that can be used on multiple platforms, just because *we* won't use that in the repository right now.

**Comment by koying** [comment]
> Would be nice to not destroy the ability to make a package that can be used on multiple platforms

Not saying, but then we need something like https://github.com/koying/xbmc/commit/54e87abeab14c28dc95aaea23ef28759be7131cc (for android)

**Comment by wsnipex** [comment]
no, we just need a way to tell the client which package to download. We can assume each client will only have exactly one - arch specific (or not) - version installed. The actual library filenames inside the packages can be the same.

I'd go with tamland's suggestion

**Comment by koying** [comment]
> I'd go with tamland's suggestion

@tamland just said he wanted to keep the possibility to have multiple arch per package ;)

The discussion starts being confusing. We might want to bullet and agree on the requirements

**Comment by Rechi** [comment]
@tamland please have a look at the path attribute implementation. I'm not sure if it is done at the place you want.

File: xbmc/addons/AddonManager.cpp
Comment (tamland) [inline xbmc/addons/AddonManager.cpp]: Not that big of a deal but can't you refactor `Factory` so this whole can be moved there? So many edge cases spread all over now. I don't think the 'info->plugin_path' hack right below is needed either anymore.

Comment (Rechi) [reply xbmc/addons/AddonManager.cpp]: @tamland please have a look at the modified implementation of path attribute

Comment (tamland) [reply xbmc/addons/AddonManager.cpp]: looks good. thanks!

**Comment by MartijnKaijser** [comment]
beta is nearing. can we get this in as we'll need it. Don't forget to change the temp repo

**Comment by Rechi** [comment]
We first have to agree which of the following two possible solutions (or maybe a complete different one) `<platform>#PLATFORM#-#ARCH#</platform>`  or `library_#PLATFORM#-#ARCH#=` to use.

**Comment by MartijnKaijser** [comment]
If no decision is made i'm going to roll a dice and that will be final

File: cmake/KodiConfig.cmake.in
Comment (wsnipex) [inline cmake/KodiConfig.cmake.in]: please move the logic out of this file (i.e. to archsetup). We just want the resulting PLATFORM_TAG here.

Comment (Rechi) [reply cmake/KodiConfig.cmake.in]: Moving it to ArchSetup.cmake doesn't work for anything except linux distros (as they use `<PREFIX>/share/kodi/cmake/KodiConfig.cmake` from an already existing Kodi install). All others generate KodiConfig.cmake during building binary add-ons, but that never includes `cmake/scripts/*/ArchSetup.cmake`.

https://github.com/xbmc/xbmc/blob/dd30a2b0723ad97ffb5944d54d6d4a850fca29fe/cmake/scripts/common/PrepareEnv.cmake#L35-L36

Comment (wsnipex) [reply cmake/KodiConfig.cmake.in]: then move it to PrepareEnv?

**Comment by garbear** [comment]
Is it possible to disable binary add-ons per-architecture? I have reports that uae4arm (an arm-specific core) doesn't build for amd64: https://github.com/garbear/xbmc/issues/108

**Comment by garbear** [comment]
doesn't that control the platform, not the architecture?

**Comment by wsnipex** [comment]
right, sorry. No arch-specific white/blacklist atm.

**Comment by MartijnKaijser** [comment]
If it doesn't succesfully build we currently automatically don't ship it so it's no so important for now

**Comment by garbear** [comment]
A user complained it interrupted their build script. But true, other things to worry about for now.

File: cmake/scripts/common/AddonHelpers.cmake
Comment (Rechi) [inline cmake/scripts/common/AddonHelpers.cmake]: Should I keep this or are addon maintainers forced to adjust their `addon.xml.in` before it can be accepted into the binary addons repo?

Comment (MartijnKaijser) [reply cmake/scripts/common/AddonHelpers.cmake]: Can we do some grace period that we support both between now and just before final so they can adjust over coming month?

Comment (Rechi) [reply cmake/scripts/common/AddonHelpers.cmake]: Yes, but we can't let the ones using the old style into the binary repo without this hack.
e.g.: old style is android, which would match `android-armv7` and `android-aarch64`

Comment (MartijnKaijser) [reply cmake/scripts/common/AddonHelpers.cmake]: It's more for transition period between shipping it with core and moving it to repo

Comment (Rechi) [reply cmake/scripts/common/AddonHelpers.cmake]: Bundled binary add-ons will always work (no hack & not updated **OR** no hack & update **OR** with the hack).

Comment (wsnipex) [reply cmake/scripts/common/AddonHelpers.cmake]: I'd say keep it for v18, but remove for v19

**Review by wsnipex — APPROVED** [review]

**Review by AlwinEsch — APPROVED** [review]

**Pull request merged**
